lib/uhash.ts

```diff
@@ -0,0 +1,54 @@
+// This module implements the "uhash" signature specification described in docs/signature_specification.md
+import * as crypto from 'crypto';
+import { SignatureResult } from './types';
+import { isBinary } from './utils/binary';
+
+const hashAlgorithm = 'md5';
+
+export const getUhashSignature = (
+  filePath: string,
+  fileContents: Buffer,
+): SignatureResult => {
+  const digest = getUhashDigest(fileContents, isBinary(fileContents));
+  const data = digest.slice(0, 24);
+  return {
+    path: filePath,
+    hashes_ffm: [
+      {
+        data,
+        format: 3,
+      },
+    ],
+  };
+};
+
+const getUhashDigest = (fileContents: Buffer, isBinary: boolean): string => {
+  const file = isBinary ? fileContents : removeWhitespaceBytewise(fileContents);
+  if (hashAlgorithm == 'md5') {
+    const hash = crypto.createHash(hashAlgorithm).update(file);
+    return hash.digest('hex');
+  } else {
+    // placeholder for non-md5 hashing algorithm, e.g. xxhash
+    return '';
+  }
+};
+
+export const removeWhitespaceBytewise = (fileBuffer: Buffer): Buffer => {
+  const BYTES_TO_REMOVE = new Uint8Array([0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x20]);
+  return removeBytesFromBuffer(fileBuffer, BYTES_TO_REMOVE);
+};
+
+const removeBytesFromBuffer = (
+  fileBuffer: Buffer,
+  valuesToRemove: Uint8Array,
+): Buffer => {
+  let writeIndex = 0;
+  for (let readIndex = 0; readIndex < fileBuffer.length; readIndex++) {
+    if (!valuesToRemove.includes(fileBuffer[readIndex])) {
+      fileBuffer[writeIndex] = fileBuffer[readIndex];
+      writeIndex++;
+    }
+  }
+  const filteredBuffer = fileBuffer.subarray(0, writeIndex);
```

Comment:
can you get rid of one of the two copies buy using fileBuffer.slice?

Reply:
I will try it. We did quite a lot of benchmarking with different implementations. I can't remember if I tried that already.

Reply:
I tried following on my machine

```
> let origin = new Int8Array([1,2,3,4,5])
> let sub = origin.subarray(0, 5)
> let slice = origin.slice(0, 5)

> sub
Int8Array(5) [ 1, 2, 3, 4, 5 ]
> slice
Int8Array(5) [ 1, 2, 3, 4, 5 ]


> slice[0] = 99
99
> slice
Int8Array(5) [ 99, 2, 3, 4, 5 ]
> origin
Int8Array(5) [ 1, 2, 3, 4, 5 ]



> sub[0] = 99
99
> sub
Int8Array(5) [ 99, 2, 3, 4, 5 ]
> origin
Int8Array(5) [ 99, 2, 3, 4, 5 ]
```

Slice would create a copy of the original, while sub would use the same underlying array. This is also described in the documentation; hence if we want to reuse the data, we should use subarray.

And we are already mutating incoming data, so it's probably the correct behavior to reuse memory with a subarray.

```
> function A(a) {
... a[1] = 77;
... }
undefined
> A(origin)
undefined
> origin
Int8Array(5) [ 99, 77, 3, 4, 5 ]
>
```

Is it correct behavior from a user's perspective? Does another developer expect the data to be "destroyed" when provided as an argument, or would that surprise the other developer? Since the function is for internal use, this might be ok, but do we need to document that behavior, since it's an unexpected implementation detail?

Reply:
I don't have a strong opinion about using either slice or subarray, but regarding reusing the data, it's hard to imagine a need to reusing the data vs. copying, I don't think there's really any danger here. If someone were to (for some unknown reason) resuse `fileContents` after the call to`removeWhiespaceBytewise`, they'd surely have to understand what `removeWhiespaceBytewise` is doing to that object.

On another topic, I did some benchmarking of slice vs. subarray, and I didn't find any consistent, measurable difference between the two implementations. For reference, here are the two implementations I benchmarked:

Slice version:
```typescript
const removeBytesFromBuffer = (
  fileBuffer: Buffer,
  valuesToRemove: Uint8Array,
): Buffer => {
  let writeIndex = 0;
  for (let readIndex = 0; readIndex < fileBuffer.length; readIndex++) {
    if (!valuesToRemove.includes(fileBuffer[readIndex])) {
      fileBuffer[writeIndex] = fileBuffer[readIndex];
      writeIndex++;
    }
  }
  return fileBuffer.slice(0, writeIndex);
};
```
Subarray version:

```typescript
const removeBytesFromBuffer3 = (
  fileBuffer: Buffer,
  valuesToRemove: Uint8Array,
): Buffer => {
  let writeIndex = 0;
  for (let readIndex = 0; readIndex < fileBuffer.length; readIndex++) {
    if (!valuesToRemove.includes(fileBuffer[readIndex])) {
      fileBuffer[writeIndex] = fileBuffer[readIndex];
      writeIndex++;
    }
  }
  const filteredBuffer = fileBuffer.subarray(0, writeIndex);
  return Buffer.from(filteredBuffer);
};
```

Reply:
> Does another developer expect the data to be "destroyed" when provided as an argument

@dekelund hey, the function itself explicit states that the goal is to get rid of WhitespaceBytewise,  
so it should not be a problem as it's very specific.

I chatted with @robguinness-snyk and we decided of move `removeBytesFromBuffer` to `removeWhitespaceBytewise` and have a single function.

Reply:
> In your case it's the opposite, since you are always modifying the target you don't need the return, and that might be away to reduce the confusion. But you're right, the name might indicate the behavior.

I think we do need the return, whether slice is used or subarray. Even though subarray is not modfiying the data in fileBuffer, we are still generating a new object that references a part of the original array, and this needs to be returned.

Reply:
> Regarding the benchmarking, that's a really hard thing to to, especially if you're running it on your local machine.
Then there is so many things that might impact, you will have side-effects from browsers, operating systems, javascript itself etc. I don't know how you ran your benchmark, but regardless I know it's hard, even the environment in the room (heating etc) might impact the benchmark.

I agree. It's hard to get benchmarking right. :-) With the level of benchmarking I did, you won't be able to measure small differences. I'm mainly interested if there are large differences between the different proposed implementations.